### DISCUSSION TITLE
Job Restart at an arbitrary block number?

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/grblHAL#242] [Ideas]
This is a core Grbl issue but there are obviously sender issues.  Maybe it could be implemented solely in a sender?

A recurring issue for many users is the problem of getting half way through a job and having to stop, for what ever reason (bit breaks, clamp problem, power fail...).  I have had this happen several times. Rather than rerun the program from start, I edit the GCode to start close to where it left off.  This requires a lot of care in understanding what the machine state should be when you begin cutting again - X/Y/Z/A/B/C position, feedrate, spindle speed, mist/cool.

Higher end CNC machines (Okuma, for example) allow you to select a restart block number.  The controller runs through the GCode up to the restart block so that all the machine settings and positions are what would been if the program had actually run up to that point. 

This would be a really nice differentiating feature in grblHAL.

**Comment by terjeio** [comment]
> Maybe it could be implemented solely in a sender?

Hmm, the sender needs to emulate from the current controller state and somehow send a new one from a dry run up to the restart point? Better to have something similar to check mode in the controller that can be exited at a given line?

**Comment by breiler** [comment]
We have tried implementing this in UGS, here is an example of the visualization changing when choosing to "Run from":
![2021-01-20 20-30-55 2021-01-20 20_39_43](https://user-images.githubusercontent.com/8962024/105226039-bd3f6000-5b5f-11eb-9d7e-01b35a105377.gif)

As you mentioned, this requires a lot of understanding of the GCode state for different machines to be able to visualize it and to start running from the given state. I'm certain that we didn't cover everything in our implementation but so far it seems to work pretty well.

The benefit of having this in the sender is that it can be used by (maybe) all supported controllers. The start will also be instantaneous as opposed to if we had to stream the program in a "check mode"-state up until a certain block.

I'm not saying it shouldn't be done in grblHAL, just that I think this feature would possibly get a bigger bang for the buck if it is implemented in the sender.

**Comment by einencool** [comment]
In the past I used the bCNC Software.
And this software has a nice feature, that you can see where you have stopped the program and then you have also a GCode Editor built in, and it shows your code in „segments“. For example you have multiple parts with different heights, it shows a list with several different parts of your code and you can just „uncheck“ the parts, you don‘t need anymore.

I found this feature absolutely great. Or better to say, I love it :-)
It also displays, what part of your code was already done with different colors.

For me the „3D-Viewer“ in IO-Sender is a bit confusing, because of the „view-angle“ but that‘s another story...

But if you take a look at the bCNC Software, that‘s a great piece of software. (In my opinion)

These two parts (different colored and „clear“ view of Code and the possibility to start in the middle of the GCode) I‘m really missing from bCNC, but your IO-Sender is getting better and better

**Reply by Klippert** [reply]
Implementing that would ba absolutely amazing! I have so many times had jobs that fails part way through, for various reasons - and the long of  the short is that you either end up going back to fusion and create a new toolpath or miling an acceptable amount of air (with possible rubbing and whatnot) to complete the job.

Big supporting cheers from me ;)

/K
